include/drivers/video.h

```diff
@@ -0,0 +1,518 @@
+
+/**
+ * @typedef video_api_stream_start_t
+ * @brief Start the capture or output process.
+ * See video_stream_start() for argument descriptions.
+ */
+typedef int (*video_api_stream_start_t)(struct device *dev);
+
+/**
+ * @typedef video_api_stream_stop_t
+ * @brief Stop the capture or output process.
+ * See video_stream_stop() for argument descriptions.
+ */
+typedef int (*video_api_stream_stop_t)(struct device *dev);
+
+/**
+ * @typedef video_api_set_ctrl_t
+ * @brief Stop the capture or output process.
+ * See video_set_ctrl() for argument descriptions.
+ */
+typedef int (*video_api_set_ctrl_t)(struct device *dev, unsigned int cid,
+				    void *value);
+
+/**
+ * @typedef video_api_get_ctrl_t
+ * @brief Stop the capture or output process.
+ * See video_get_ctrl() for argument descriptions.
+ */
+typedef int (*video_api_get_ctrl_t)(struct device *dev, unsigned int cid,
+				    void *value);
+
+/**
+ * @typedef video_api_get_caps_t
+ * @brief Get capabilities of a video endpoint.
+ * See video_get_caps() for argument descriptions.
+ */
+typedef int (*video_api_get_caps_t)(struct device *dev,
+				    enum video_endpoint_id ep,
+				    struct video_caps *caps);
+
+/**
+ * @typedef video_event_cb
+ * @brief Video endpoint callback prototype.
+ *
+ * @param dev Pointer to the device structure.
+ * @param ep Endpoint ID.
+ * @param event Event type.
+ * @param user_data Pointer to user data.
+ *
+ */
+typedef void (*video_event_cb)(struct device *dev, enum video_endpoint_id ep,
+			       enum video_event event,
+			       void *user_data);
+
+/**
+ * @typedef video_api_register_event_cb_t
+ * @brief Register callback for video endpoint events.
+ * See video_register_event_cb() for argument descriptions.
+ */
+typedef int (*video_api_register_event_cb_t)(struct device *dev,
+					     enum video_endpoint_id ep,
+					     video_event_cb cb,
+					     void *user_data);
+
+struct video_driver_api {
+	video_api_set_format_t set_format;
+	video_api_get_format_t get_format;
+	video_api_enqueue_t enqueue;
+	video_api_dequeue_t dequeue;
+	video_api_stream_start_t stream_start;
+	video_api_stream_stop_t stream_stop;
+	video_api_set_ctrl_t set_ctrl;
+	video_api_set_ctrl_t get_ctrl;
+	video_api_get_caps_t get_caps;
+	video_api_register_event_cb_t register_event_cb;
+};
+
+/**
+ * @brief Set video format.
+ *
+ * Configure video device with a specific format
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param ep Endpoint ID.
+ * @param fmt Pointer to a video format struct.
+ *
+ * @retval 0 Is successful.
+ * @retval -EINVAL If parameters are invalid.
+ * @retval -ENOTSUP If format is not supported.
+ * @retval -EIO General input / output error.
+ */
+static inline int video_set_format(struct device *dev,
+				   enum video_endpoint_id ep,
+				   struct video_format *fmt)
+{
+	const struct video_driver_api *api =
+		(const struct video_driver_api *)dev->driver_api;
+
```

Comment:
IMHO we should have ASSERTs checking if the backing API is implemented (for mandatory methods) or we should return -ENOSUP (if method is optional and not implemented)

Reply:
OK, I'm going to add that and clarify which one is mandatory/optional